src/rule_redirects.rs

```diff
@@ -0,0 +1,153 @@
+use std::collections::HashMap;
+use std::str::FromStr;
+
+use once_cell::sync::Lazy;
+
+use crate::registry::RuleSelector;
+
+pub fn get_rule_redirect(code: &str) -> Option<&str> {
+    REDIRECTS.get(code).copied()
+}
+
+#[derive(Debug, Clone)]
+pub struct MaybeRedirectedRuleSelector(RuleSelector);
+
+impl FromStr for MaybeRedirectedRuleSelector {
+    type Err = <RuleSelector as FromStr>::Err;
+
+    fn from_str(code: &str) -> Result<Self, Self::Err> {
+        Ok(Self(RuleSelector::from_str(
+            match get_rule_redirect(code) {
+                Some(redirect_target) => {
+                    crate::warn_user_once!("`{code}` has been remapped to `{redirect_target}`");
```

Comment:
Does this correctly trigger once, _per rule_? Or is this going to trigger once, ever?

Reply:
Hahaha oh yeah ... this indeed didn't work as intended.

I just changed it to just use `warn_user!` since I cannot think of an easy alternative to preserve the limitation that it only prints once since a `FromStr` impl cannot easily keep state.

Reply:
Hmm... It'd be nice to preserve, although it's not the end of the world. It looks like it appears once per inclusion, and never when parsing `noqa` redirects (haven't looked at the code to understand how / why the latter behavior works -- does that make sense to you?).

Given `foo.py`:

```py
import pandas as x # noqa: IC001
```

I get:

```
❯ cargo run foo.py --select IC001 --fixable IC001
    Finished dev [unoptimized + debuginfo] target(s) in 0.13s
     Running `target/debug/ruff foo.py --select IC001 --fixable IC001`
warning: `IC001` has been remapped to `ICN001`
warning: `IC001` has been remapped to `ICN001`
warning: debug build without --no-cache.
```

Reply:
(Oh, I guess it's because we generate the redirect map, then accesses on it don't trigger the warning.)

Reply:
Oh, right, ok, it's because those go through the `REDIRECTS` map directly.

Reply:
ruff didn't warn about redirects in `noqa` comments previously ... my PR doesn't change that.

Anyway I have updated the PR to only warn about redirects once when they occur multiple times (in the CLI or `pyproject.toml` files) :)